### PR TITLE
Notebooks: Grid Support

### PR DESCRIPTION
**Pull request by chlafreniere**
Creating a new renderer for the 'application/vnd.dataresource+json' MIME type. @alanrenmsft suggested that I take a look at TableDataView, so leveraging that alongside Table to create a grid for our notebook cell outputs.

For the SQL kernel, we will see the new grid. Notebooks saved in ADS can be opened/viewed in the Jupyter viewer. I also ensured that we correctly support 'application/vnd.dataresource+json' by opening an existing notebook saved in another viewer; the table and its data displayed as expected.

### CODE REVIEW

File: src/sql/parts/notebook/outputs/factories.ts [hunk 1]
```diff
@@ -80,6 +80,16 @@ export const javaScriptRendererFactory: IRenderMime.IRendererFactory = {
 	createRenderer: options => new widgets.RenderedJavaScript(options)
 };
 
+export const dataResourceRendererFactory: IRenderMime.IRendererFactory = {
+	safe: true,
+	mimeTypes: [
+		'application/vnd.dataresource+json',
+		'application/vnd.dataresource'
+	],
+	defaultRank: 40,
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/factories.ts]: Ask: check JupyterLab & NTeract's renderer priority and see if they've changed / how they prioritize these.

File: src/sql/parts/notebook/outputs/renderers.ts [hunk 2]
```diff
@@ -265,6 +271,88 @@ export function renderLatex(
...
 }
 
+/**
+ * Render DataResource as a grid into a host node.
+ *
+ * @params options - The options for rendering.
+ *
+ * @returns A promise which resolves when rendering is complete.
+ */
+export function renderDataResource(
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/renderers.ts]: Can we put this to its own file? I think we'll want to do a lot more here, so would be good to have it's own place where we can put lots of logic

File: src/sql/parts/notebook/outputs/renderers.ts [hunk 3]
```diff
@@ -265,6 +271,88 @@ export function renderLatex(
...
+	options: renderDataResource.IRenderOptions
+): Promise<void> {
+	// Unpack the options.
+	let { host, source } = options;
+	let sourceObject: IDataResource = JSON.parse(source);
+
+	// Before doing anything, avoid re-rendering the table multiple
+	// times (as can be the case when going untrusted -> trusted)
+	while (host.firstChild) {
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/renderers.ts]: If you've scrolled down the grid and change trusted -> untrusted is this going to reset so we go to the top of the results? Should we have some way of avoiding re-rendering unnecessarily?

File: src/sql/parts/notebook/outputs/renderers.ts [hunk 4]
```diff
@@ -265,6 +271,88 @@ export function renderLatex(
...
+		host.removeChild(host.firstChild);
+	}
+
+	// now create the table
+	let tableContainer = document.createElement('div');
+	tableContainer.className = 'notebook-cellTable';
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/renderers.ts]: Will we ensure CSS matches other grids?

File: src/sql/parts/notebook/outputs/renderers.ts [hunk 5]
```diff
@@ -265,6 +271,88 @@ export function renderLatex(
...
+
+	const ROW_HEIGHT = 29;
+	const BOTTOM_PADDING_AND_SCROLLBAR = 14;
+	let tableResultsData = new TableDataView();
+	let columns: string[] = sourceObject.schema.fields.map(val => val.name);
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/renderers.ts]: Should we validate schema has correct number of entries vs. data? Some validation should be important here, let's not assume it's correct coming in.

File: src/sql/parts/notebook/outputs/renderers.ts [hunk 6]
```diff
@@ -265,6 +271,88 @@ export function renderLatex(
...
+	// In order to show row numbers, we need to put the row number column
+	// ahead of all of the other columns, and register the plugin below
+	let rowNumberColumn = new RowNumberColumn({ numberOfRows: source.length });
+	columnsTransformed.unshift(rowNumberColumn.getColumnDefinition());
+
+	tableResultsData.push(transformData(sourceObject.data, columns));
+
+	let detailTable = new Table(tableContainer, {
+		dataProvider: tableResultsData, columns: columnsTransformed
+	}, { rowHeight: ROW_HEIGHT,
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/renderers.ts]: Minor: please put rowHeight on next line

File: src/sql/parts/notebook/outputs/renderers.ts [hunk 7]
```diff
@@ -265,6 +271,88 @@ export function renderLatex(
...
+		forceFitColumns: false,
+		defaultColumnWidth: 120
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/renderers.ts]: Where are the defaults coming from?

File: src/sql/parts/notebook/outputs/renderers.ts [hunk 8]
```diff
@@ -265,6 +271,88 @@ export function renderLatex(
...
+	 });
+	detailTable.setSelectionModel(new CellSelectionModel());
+	detailTable.registerPlugin(rowNumberColumn);
+
+	// Need to include column headers and scrollbar, so that's why 1 needs to be added
+	let rowsHeight = (detailTable.grid.getDataLength() + 1) * ROW_HEIGHT + BOTTOM_PADDING_AND_SCROLLBAR;
+
+	// Set the height dynamically if the grid's height is < 500px high; otherwise, set height to 500px
+	tableContainer.style.height = rowsHeight >= 500 ? '500px' : rowsHeight.toString() + 'px';
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/renderers.ts]: Minor: consider a setting for this somewhere so we can find it & change easier.

File: src/sql/parts/notebook/outputs/renderers.ts [hunk 9]
```diff
@@ -265,6 +271,88 @@ export function renderLatex(
...
+	let columns: string[] = sourceObject.schema.fields.map(val => val.name);
+	// Table object requires passed in columns to be of datatype Slick.Column
+	let columnsTransformed = transformColumns(columns);
+
+	// In order to show row numbers, we need to put the row number column
+	// ahead of all of the other columns, and register the plugin below
+	let rowNumberColumn = new RowNumberColumn({ numberOfRows: source.length });
+	columnsTransformed.unshift(rowNumberColumn.getColumnDefinition());
+
+	tableResultsData.push(transformData(sourceObject.data, columns));
```
Comment (kevcunnane) [inline src/sql/parts/notebook/outputs/renderers.ts]: Can you check with Alan & core team on whether TableDataView handles cross-site scripting protection of the data we input? We've had reports (https://medium.com/dealeron-dev/electron-and-xss-e9ecef1c7325) where with HTML results we effectively XSS ourselves. I want to ensure we don't make the same mistake in Notebooks too.

Comment (kevcunnane) [reply src/sql/parts/notebook/outputs/renderers.ts]: Note you can test this with HTML content - should have existing bug examples in the recent fixes for this.

File: src/sql/workbench/services/notebook/common/sqlSessionManager.ts [hunk 10]
```diff
@@ -299,7 +299,24 @@ export class SQLFuture extends Disposable implements FutureInternal {
 	setIOPubHandler(handler: nb.MessageHandler<nb.IIOPubMessage>): void {
 		this._register(this._queryRunner.onBatchEnd(batch => {
 			this._queryRunner.getQueryRows(0, batch.resultSetSummaries[0].rowCount, 0, 0).then(d => {
-				let data: SQLData = {
+				let columns = batch.resultSetSummaries[0].columnInfo.map(c => c.columnName);
```
Comment (kevcunnane) [inline src/sql/workbench/services/notebook/common/sqlSessionManager.ts]: Please refactor to 2 functions `convertToDataResource` and `convertToHtmlTable`. We may want to use these elsewhere, plus it's easier to understand example what each part of this does.

Comment (kevcunnane) [reply src/sql/workbench/services/notebook/common/sqlSessionManager.ts]: Also maybe a stretch but... unit tests??? 😄

File: src/sql/parts/notebook/outputs/renderers.ts (hunk 7 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/outputs/renderers.ts]: They match the query editor's values (which unfortunately are hardcoded as well). Any ideas where I can put this in a common place?

Comment (kevcunnane) [reply src/sql/parts/notebook/outputs/renderers.ts]: @anthonydresser can you advise?

File: src/sql/parts/notebook/outputs/renderers.ts (hunk 9 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/outputs/renderers.ts]: Good catch. I've fixed these issues with XSS'ing ourselves here. Will work with team to get a more generic solution going.

File: src/sql/parts/notebook/outputs/renderers.ts (hunk 5 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/outputs/renderers.ts]: If it's not, what should we do?

File: src/sql/parts/notebook/outputs/factories.ts (hunk 1 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/outputs/factories.ts]: I found the code for JupyterLab, but not for NTeract. Can you point me in the right direction there?

The numbers from JupyterLab are different, but it follows the same structure, where specific renderers have a rank that's lower than more generic renderers.

File: src/sql/parts/notebook/outputs/renderers.ts (hunk 3 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/outputs/renderers.ts]: Not from what I could see; we'd need to differentiate these cases, and that's nontrivial :(.

File: src/sql/parts/notebook/outputs/renderers.ts (hunk 4 above)
Comment (chlafreniere) [reply src/sql/parts/notebook/outputs/renderers.ts]: We already do share a lot of CSS

**Pull request merged**